port/cpl_vsil_curl.cpp

```diff
@@ -4880,6 +4893,84 @@ int VSICurlParseUnixPermissions(const char* pszPermissions)
     return nMode;
 }
 
+/************************************************************************/
+/*                  Cache of file properties.                           */
+/************************************************************************/
+
+static std::mutex oCacheFilePropMutex;
+static lru11::Cache<std::string, FileProp>* poCacheFileProp = nullptr;
+
+/************************************************************************/
+/*                   VSICURLGetCachedFileProp()                         */
+/************************************************************************/
+
+bool VSICURLGetCachedFileProp( const char* pszURL, FileProp& oFileProp )
+{
+    std::lock_guard<std::mutex> oLock(oCacheFilePropMutex);
+    return poCacheFileProp != nullptr &&
+           poCacheFileProp->tryGet(std::string(pszURL), oFileProp) &&
+            // Let a chance to use new auth parameters
+           !(oFileProp.eExists == EXIST_NO &&
+             gnGenerationAuthParameters != oFileProp.nGenerationAuthParameters);
+}
+
+/************************************************************************/
+/*                   VSICURLSetCachedFileProp()                         */
+/************************************************************************/
+
+void VSICURLSetCachedFileProp( const char* pszURL, FileProp& oFileProp )
+{
+    std::lock_guard<std::mutex> oLock(oCacheFilePropMutex);
+    if( poCacheFileProp == nullptr )
+        poCacheFileProp = new lru11::Cache<std::string, FileProp>(100 * 1024);
+    oFileProp.nGenerationAuthParameters = gnGenerationAuthParameters;
+    poCacheFileProp->insert(std::string(pszURL), oFileProp);
+}
+
+/************************************************************************/
+/*                   VSICURLInvalidateCachedFileProp()                  */
+/************************************************************************/
+
+void VSICURLInvalidateCachedFileProp( const char* pszURL )
+{
+    std::lock_guard<std::mutex> oLock(oCacheFilePropMutex);
+    if( poCacheFileProp != nullptr )
+        poCacheFileProp->remove(std::string(pszURL));
+}
+
+/************************************************************************/
+/*               VSICURLInvalidateCachedFilePropPrefix()                */
+/************************************************************************/
+
+void VSICURLInvalidateCachedFilePropPrefix( const char* pszURL )
+{
+    std::lock_guard<std::mutex> oLock(oCacheFilePropMutex);
+    if( poCacheFileProp != nullptr )
+    {
+        std::list<std::string> keysToRemove;
```

Comment:
why are you using a list here instead of vector?

Reply:
I guess both could make sense (we just need iterative access, not random access, so std::list is not inappropriate). The lru11::Cache implementation in port/cpl_mem_cache.h (that is a third party originally) uses a std::list internally too